strapi/strapi
# [Upgrade Tool] V2

### What does it do?

- Modified the CLI usage to match new criterias
	- `upgrade` Don't do anything and display the help
	- `upgrade major` Will try upgrading to the next available major (only if the project is already on the latest minor.patch version for the current major), if none is found, it'll abort
	- `upgrade minor` _todo in another PR_
	- `upgrade patch` _todo in another PR_
- Refactored the overall structure of the `upgrade` package
	- [src/cli/index.ts](packages/utils/upgrade/src/cli/index.ts) contains the overall definition of the CLI usage
	- `src/cli/commands` contains actual commands' definitions, including the one for the [upgrade](packages/utils/upgrade/src/cli/commands/upgrade.ts)
	- The [upgrade task](packages/utils/upgrade/src/tasks/upgrade.ts) is defined in the `src/tasks` directory and is directly called by the upgrade command
	- Every domain-related piece of logic has been split as individual modules available at `src/modules/*`
		- Each module has a consistent folder structure that can contain:
			- `types.ts` exports types related to the module
			-  `constants.ts` exports constants related to the module
			- `xyz.ts` contains domain logic related to the module (usually it exports a factory to create related domain objects. e.g: `xyzFactory()`)
			- `index.ts` is considered the public module API and only exports what can be used by anyone outside the module
			- A `__tests__/` directory which contains domain-related tests
- Added unit tests
- Fixed lint & build issues
- Bumped `fs-extra` from `10.0.0` to `10.1.0` to fix compatibility issues with `memfs` which is used for mocking `fs` during tests (see [fs-extra's changelog](https://github.com/jprichardson/node-fs-extra/blob/10.1.0/CHANGELOG.md#1010--2022-04-16) and related [issue](https://github.com/streamich/memfs/issues/803) / [fix PR](https://github.com/jprichardson/node-fs-extra/pull/953))

### Why is it needed?

After discussing at length the scope of the upgrade tool internally, we've decided to modify it a bit (mainly related to available commands, options, and how they should behave)

This PR aims to refactor the tool to incorporate those changes while cleaning up the overall architecture.

### How to test it?

- Make sure your monorepository is up-to-date (install, setup, etc...)
- From the `examples/getstarted` directory, use `../../packages/utils/upgrade/bin/upgrade.js <command> [...options]` to test the upgrade CLI
